glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,543 @@
+        }
+    }
+
+    #[test]
+    fn executor_stall_detector() {
+        let stall_handler = TestHandler::new(nix::libc::SIGUSR1 as u8);
+        LocalExecutorBuilder::default()
+            .detect_stalls(Some(Box::new(stall_handler.clone())))
+            .preempt_timer(Duration::from_millis(50))
+            .make()
+            .unwrap()
+            .run(async {
+                // will trigger the stall detector because we go over budget
+                thread::sleep(Duration::from_millis(100));
+
+                let exec = crate::executor();
+                assert!(stall_handler.inner.read().unwrap().detections.is_empty());
+
+                exec.yield_task_queue_now().await; // yield the queue
+
+                assert!(
+                    stall_handler
+                        .inner
+                        .write()
+                        .unwrap()
+                        .detections
+                        .pop()
+                        .is_some()
+                );
+
+                // no stall because < 50ms of un-cooperativeness
+                thread::sleep(Duration::from_millis(40));
+
+                assert!(stall_handler.inner.read().unwrap().detections.is_empty());
+
+                exec.yield_task_queue_now().await; // yield the queue
+
+                // no stall because a timer yields internally
+                sleep(Duration::from_millis(100)).await;
+
+                exec.yield_task_queue_now().await; // yield the queue
+
+                assert!(stall_handler.inner.read().unwrap().detections.is_empty());
+
+                // trigger one last time
+                thread::sleep(Duration::from_millis(100));
+
+                exec.yield_task_queue_now().await; // yield the queue
+
+                assert!(
+                    stall_handler
+                        .inner
+                        .write()
+                        .unwrap()
+                        .detections
+                        .pop()
+                        .is_some()
+                );
+
+                // Make sure nothing else was reported
+                exec.yield_task_queue_now().await; // yield the queue
+                assert!(stall_handler.inner.read().unwrap().detections.is_empty());
+            });
+    }
+
+    #[test]
+    fn stall_detector_correct_signal_handler() {
+        let mut build_handlers: Vec<(TestHandler, LocalExecutorBuilder)> = Vec::with_capacity(10);
+        for i in 1..11 {
+            let handler = TestHandler::new(nix::libc::SIGUSR1 as u8);
+            let tname = format!("exec{}", i);
+            let builder = LocalExecutorBuilder::default()
+                .name(&tname)
+                .detect_stalls(Some(Box::new(handler.clone())))
+                .preempt_timer(Duration::from_millis(50));
+            build_handlers.push((handler, builder));
+        }
+        let mut handles = Vec::with_capacity(10);
+        for (handler, builder) in build_handlers.drain(..) {
```

Comment:
Not an issue but FYI you don't need to drain here, just iterate on `build_handlers` and that will consume it.